### PR TITLE
wheel: support for 'plugin' type entry_points

### PR DESCRIPTION
**Pull request by dhalperi**
## PR Checklist

- [X] Does not include precompiled binaries, eg. `.par` files. See [CONTRIBUTING.md](/CONTRIBUTING.md) for info
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [X] Feature (please, look at the "Scope of the project" section in the README.md file)

## What is the current behavior?

Python wheels have a rich concept called `entry_points`. The most common use of these is the existing `console_scripts` support: command-line wrappers for Python code.

However, there are many more uses. Another common one is `plugins` -- the Python version of, say, Java `@AutoService`. This is how `pytest` finds modules on the `PYTHONPATH`.

This PR is something we've been using for a while to add our own plugins to the `entry_points` file; but the change is worth discussing.

In one sense, this is stupid. There's some interference between `console_scripts` and `plugins` -- in reality, the underlying structure is:

```
[group name, like console_scripts]
key1 = value1
key2 = value2
...

[g2]
key3 = v3
key4 = v4
...

...
```

So maybe this would be best as a single argument merging `console_scripts` in: `{group: {k1: v1, k2: v2, ...}, ...}`... if we wanted to make a breaking change.

But since `console_scripts` already existed, I added `plugins` separately. (I did not call it `entry_points` since it doesn't handle `console_scripts`).

(There are of course, other backwards-compat options like adding an `entry_points` arg that throws if `console_scripts` is present.)

Issue Number: not yet, but happy to file one if needed

## What is the new behavior?

Wheels built by `rules_python` support `entry_points` other than `console_scripts`, hopefully without breaking existing `console_scripts` support.

## Does this PR introduce a breaking change?

- [X] No - At least, I tried not to.

## Other information

See https://packaging.python.org/specifications/entry-points/ and specifically https://packaging.python.org/guides/creating-and-discovering-plugins/#using-package-metadata

Sorry for not ticking all the boxes from jump - this has been sitting around a while and when I realized PRs are being reviewed again I thought it might be worth starting the discussion.

---
This change is [Reviewable](https://reviewable.io/reviews/bazelbuild/rules_python/349)

### CODE REVIEW

File: experimental/python/wheel.bzl
Comment (pstradomski) [inline experimental/python/wheel.bzl:217]: That should be string_list_dict to allow more than one entry point in single group, right?

File: experimental/rules_python/wheelmaker.py
Comment (pstradomski) [inline experimental/rules_python/wheelmaker.py:161]: I'd change it so that the .bzl file just merges console_scripts and plugins, and wheelmaker just always sees only entry_points.

File: experimental/python/wheel.bzl
Comment (pstradomski) [inline experimental/python/wheel.bzl:217]: I think entry_points would be a better name.

File: experimental/python/wheel.bzl
Comment (pstradomski) [inline experimental/python/wheel.bzl:212]: The .bzl rule can just merge console_scripts into entry_points/plugins (i.e. treat entries in "console_scripts" as if they were entries in "entry_points" in the "console_script" group). I guess the question is whether console_scripts should be marked as deprecated or do we want to live with this merging forever. I'm fine with marking as deprecated.

File: experimental/rules_python/wheelmaker.py
Comment (pstradomski) [inline experimental/rules_python/wheelmaker.py:174]: This can produce multiple groups with the same name. Even if that's allowed by the spec (I'm not sure?) it would be better to order the entries so that each group appears only once.

**Review by pstradomski — CHANGES_REQUESTED** [review]

File: experimental/python/wheel.bzl
Comment (dhalperi) [reply experimental/python/wheel.bzl]: Implemented.

I documented `console_scripts` as deprecated based on how other things in this project are. I didn't see a specific Starlark flag to make an attr as deprecated.

File: experimental/python/wheel.bzl
Comment (dhalperi) [reply experimental/python/wheel.bzl]: Yep.

Comment (dhalperi) [reply experimental/python/wheel.bzl]: Done.

File: experimental/rules_python/wheelmaker.py
Comment (dhalperi) [reply experimental/rules_python/wheelmaker.py]: Done.

File: experimental/rules_python/wheelmaker.py
Comment (dhalperi) [reply experimental/rules_python/wheelmaker.py]: This did not happen before (at least from `wheel.bzl`) because there was only one entry allowed per group. As part of switching that to allow multiple entry points per group, rewrite this code from scratch. I believe all such concerns are now handled.

File: experimental/rules_python/wheelmaker.py
Comment (dhalperi) [inline experimental/rules_python/wheelmaker.py:165]: FYI, an earlier draft of this code required `ref` to be in the format `a:b`. Turns out that's only one of the allowed formats for an object reference, so I dropped it.

File: experimental/rules_python/wheelmaker.py
Comment (dhalperi) [inline experimental/rules_python/wheelmaker.py:180]: Is there a stock formatter for this repo? I tried `yapf` but it changed way more than just the modified code.

File: experimental/python/wheel.bzl
Comment (pstradomski) [reply experimental/python/wheel.bzl:215]: I think the best we can do is print("console_scripts is deprecated, please use entry_points instead") in the skylark rule when it sees non-empty console_scripts.

File: experimental/python/wheel.bzl
Comment (pstradomski) [inline experimental/python/wheel.bzl:150]: Perhaps it would be better to generate entry_points.txt in .bzl (like _target_wrapped_inputs.txt does above) and just package it in wheelmaker.py.

I leave it up to you to decide which approach you prefer.

**Review by pstradomski — CHANGES_REQUESTED** [review]
@andyscott @thundergolfer : This change looks good to me (except for the comments I've left). Should I merge it?

I authored the initial implementation of py_wheel for this repo and kept maintianing it, but with you as new maintainers I'd like a clarification if I should continue merging changes here or would you prefer to review/merge all new PRs?

File: experimental/python/wheel.bzl
Comment (dhalperi) [reply experimental/python/wheel.bzl:233]: Done.

File: experimental/python/wheel.bzl
Comment (dhalperi) [reply experimental/python/wheel.bzl]: Rewrote entirely. The ugliness moved, but is not gone. The wheelmaker.py file is much cleaner though.

File: experimental/python/wheel.bzl
Comment (dhalperi) [reply experimental/python/wheel.bzl:233]: Hmm, this `print` statement fails `buildifier`. Is there an alternative?

**Review by pstradomski — CHANGES_REQUESTED** [review]
Presubmit complains about the print() function.

Let's remove it.

**Review by pstradomski — APPROVED** [review]
Looks good to me

**Pull request merged**
